src/box/lua/config/applier/credentials.lua

```diff
@@ -525,17 +525,93 @@ end
 local privileges_action_f = function(grant_or_revoke, role_or_user, name, privs,
                                      obj_type, obj_name)
     assert(grant_or_revoke == 'grant' or grant_or_revoke == 'revoke')
+    privs = table.copy(privs)
+    table.sort(privs)
     privs = table.concat(privs, ',')
 
     -- Try to apply the action immediately. If the object doesn't exist,
     -- the sync will be applied inside the trigger on object creation/rename.
     local ok, err = pcall(box.schema[role_or_user][grant_or_revoke],
                           name, privs, obj_type, obj_name)
 
+    -- This key identifies an alert to be dropped on a successful
+    -- operation and to be set on an error.
+    --
+    -- Let's consider the following situation: a space does not
+    -- exist, but the configuration contains privileges for it
+    -- (for example 'read' and 'write'). The warning will be
+    -- issued with the "box.schema.user.grant(<...>, 'read,write',
+    -- <...>)" key.
+    --
+    -- Now, let's assume that the space is created. We should drop
+    -- the warning and, so, we should generate exactly same key to
+    -- pass to the :drop() aboard method.
+    --
+    -- 1. Is it possible that the key to drop contains just 'read',
+    --    not 'read,write'? Seems unlikely.
+    --
+    --    If the configuration was changed before the space
+    --    creation, then the old ('read,write') alert is dropped
+    --    together with all the alerts from the old configuration
+    --    and the new alert ('read') is issued on applying the new
+    --    configuration.
+    --
+    --    OTOH, if a space creation is in the same transaction
+    --    with granting a 'write' permission, it seems possible.
+    --    However, it looks unlikely that permissions for the same
+    --    user and the same object (space/function/sequence) are
+    --    granted using box.schema.user.grant() and using the
+    --    declarative configuration at the same time.
+    --
+    -- 2. Is it possible that separate calls with 'read' and
+    --    'write' will be performed? Seems, no.
+    --
+    --    privileges_from_config() glues all the config privileges
+    --    into a mapping, so different configuration forms do not
+    --    affect the result. The following two configurations are
+    --    equivalent.
+    --
+    --    credentials:
+    --      users:
+    --        myuser:
+    --          privileges:
+    --          - permissions: [read, write]
+    --            spaces: [s]
+    --
+    --    credentials:
+    --      users:
+    --        myuser:
+    --          privileges:
+    --          - permissions: [read]
+    --            spaces: [s]
+    --          - permissions: [write]
+    --            spaces: [s]
```

Comment:
Nit:

Maybe skip `credentials: users: myuser:`, the comment is long enough and we don't really need a full config to understand what is going on. Feel free to ignore.

Reply:
I would keep it with this redundancy to bring a proper context to an occasional reader.

Reply:
Ok.